### PR TITLE
[shelljs] Added construct and function signatures for shelljs/ShellString

### PR DESCRIPTION
**Pull request by msanguineti**
Please refer to #34614

I had to re-submit the code since I've botched it by rebasing my branch... Note to self: walk away from computing machinery when half asleep.

The signatures are both function and construct. They take only 1 parameter (either a `string` or a `string[]`) to comply with the docs and not to mess with the private API.

@gkalpak  sorry again for the mishap.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/shelljs/shelljs#shellstringstr

### CODE REVIEW

File: types/shelljs/shelljs-tests.ts [hunk 1]
```diff
@@ -89,24 +89,24 @@ const testPath = shell.env["path"];
 const version = shell.exec("node --version").stdout;
 
 // $ExpectType ShellString
-const version2 = shell.exec("node --version", {async: false});
+const version2 = shell.exec("node --version", { async: false });
```
Comment (ExE-Boss) [inline types/shelljs/shelljs-tests.ts]: I&nbsp;would prefer if&nbsp;you didn’t reformat the&nbsp;tests in&nbsp;this&nbsp;PR as&nbsp;it&nbsp;adds too&nbsp;much unnecessary noise.

File: types/shelljs/index.d.ts [hunk 2]
```diff
@@ -965,6 +966,51 @@ export type ShellString = string & ShellReturnValue;
 
 export type ShellArray = string[] & ShellReturnValue;
 
+export const ShellString: ShellStringFunction;
+
+export interface ShellStringFunction {
```
Comment (ExE-Boss) [inline types/shelljs/index.d.ts]: This should be&nbsp;called `ShellStringConstructor` to&nbsp;match [TypeScript’s standard&nbsp;library](https://github.com/Microsoft/TypeScript/blob/42302705f6d4266778324d8282ed743ffa305d33/lib/lib.es5.d.ts#L517-L527):
```suggestion
export interface ShellStringConstructor {
```

File: types/shelljs/index.d.ts [hunk 3]
```diff
@@ -965,6 +966,51 @@ export type ShellString = string & ShellReturnValue;
 
 export type ShellArray = string[] & ShellReturnValue;
 
+export const ShellString: ShellStringFunction;
```
Comment (ExE-Boss) [inline types/shelljs/index.d.ts]: I&nbsp;prefer putting the&nbsp;<code>export&nbsp;const</code> declaration at&nbsp;the&nbsp;end:
```suggestion
```

File: types/shelljs/index.d.ts [hunk 4]
```diff
@@ -965,6 +966,51 @@ export type ShellString = string & ShellReturnValue;
...
+	 * This type also exposes the corresponding command's stdout, stderr, and return status
+	 * code via the `.stdout` (string), `.stderr` (string), and `.code` (number) properties
+	 * respectively.
+	 *
+	 * @param value 	The string value to wrap.
+	 * @return				A string-like object with special methods.
+	 */
+	(value: string): ShellString;
+	(value: string[]): ShellArray;
+}
```
Comment (ExE-Boss) [inline types/shelljs/index.d.ts]: I&nbsp;prefer putting the&nbsp;<code>export&nbsp;const</code> declaration after the&nbsp;interface declaration, to&nbsp;match&nbsp;the&nbsp;rest&nbsp;of&nbsp;this type&nbsp;definition and&nbsp;[TypeScript’s&nbsp;standard library](https://github.com/Microsoft/TypeScript/blob/42302705f6d4266778324d8282ed743ffa305d33/lib/lib.es5.d.ts#L517-L527):
```suggestion
}

export const ShellString: ShellStringConstructor;
```

**Review by ExE-Boss — CHANGES_REQUESTED** [review]

File: types/shelljs/shelljs-tests.ts (hunk 1 above)
Comment (msanguineti) [reply types/shelljs/shelljs-tests.ts]: Sorry about that. I guess now I'll leave it like this or do you want me to revert those changes? I could commit these changes separately from the others you suggest so to minimise the overall noise and to get it back the way it once was. Let me know.

Comment (msanguineti) [reply types/shelljs/shelljs-tests.ts]: Just went ahead and reverted the formatting. Noise cancellation: on/~~off~~

**Review by ExE-Boss — APPROVED** [review]
LGTM, r+

**Comment by ExE-Boss** [comment]
I&nbsp;believe that @gkalpak should probably review this too.

**Review by uniqueiniquity — CHANGES_REQUESTED** [review]
Marking as changes requested until @gkalpak reviews as requested.
However, I'm happy to merge this, so let me know if I shouldn't wait.

**Review by gkalpak — APPROVED** [review]
:ok:

File: types/shelljs/shelljs-tests.ts (hunk 1 above)
Comment (gkalpak) [reply types/shelljs/shelljs-tests.ts]: Wow! Given #34364, I was sure you would gonna love the whitespace changes, @ExE-Boss :stuck_out_tongue:

**Review by uniqueiniquity — APPROVED** [review]

**Comment by uniqueiniquity** [comment]
Congratulations on your first DefinitelyTyped contribution!
Thank you for being a part of the community!

File: types/shelljs/shelljs-tests.ts (hunk 1 above)
Comment (ExE-Boss) [reply types/shelljs/shelljs-tests.ts]: Well, the difference between this and #34364 is that in #34364, it was gonna be a ~90% diff either way, so I decided to just convert it to tabs.

This is a PR that affects a single thing, so reformatting the tests is more or less out of scope.

**Pull request merged**